### PR TITLE
rtmp-services: Add Dacast

### PR DESCRIPTION
**Pull request by AbAb1l**
### Description

Add Dacast streaming service to OBS dropdown options.
Add plugin to get mutable RTMP URL, stream key, username, password, and encoder options from API.

### Motivation and Context

Dacast is a live streaming platform. This change allows users to stream without complicated operations.

### How Has This Been Tested?

I confirmed that it works with the Live server. I checked on Mac and Windows.
Details are shown below.

1. Log in to https://www.dacast.com/
2. Select Live Streams
3. Create Live Stream
4. Add title and select your region
5. Select Encoder Setup
6. Select Generate
7. Copy Encoder Key
8. Open OBS streaming setting
9. Select Dacast service from the dropdown list
10. Paste Encoder Key in the Encoder Key field
11. Start streaming.

### Types of changes

 New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

### CODE REVIEW

File: UI/window-basic-auto-config.cpp [hunk 1]
```diff
@@ -633,6 +633,14 @@ void AutoConfigStreamPage::UpdateKeyLink()
...
 	}
 
+	if (serviceName == "Dacast") {
+		ui->streamKeyLabel->setText(
+			QTStr("Basic.AutoConfig.StreamPage.EncoderKey"));
+	} else {
+		ui->streamKeyLabel->setText(
+			QTStr("Basic.AutoConfig.StreamPage.StreamKey"));
+	}
+
```
Comment (notr1ch) [inline UI/window-basic-auto-config.cpp:643]: You shouldn't need to add a new translation and a hard coded service check like this. Just go with stream key.

File: UI/window-basic-settings-stream.cpp [hunk 2]
```diff
@@ -290,6 +290,14 @@ void OBSBasicSettings::UpdateKeyLink()
...
 	}
 
+	if (serviceName == "Dacast") {
+		ui->streamKeyLabel->setText(
+			QTStr("Basic.AutoConfig.StreamPage.EncoderKey"));
+	} else {
+		ui->streamKeyLabel->setText(
+			QTStr("Basic.AutoConfig.StreamPage.StreamKey"));
+	}
+
```
Comment (notr1ch) [inline UI/window-basic-settings-stream.cpp:300]: You shouldn't need to add a new translation and a hard coded service check like this. Just go with stream key.

File: deps/file-updater/file-updater/file-updater.c [hunk 3]
```diff
@@ -131,6 +131,8 @@ static bool do_http_request(struct update_info *info, const char *url,
 	// A lot of servers don't yet support ALPN
 	curl_easy_setopt(info->curl, CURLOPT_SSL_ENABLE_ALPN, 0);
 #endif
+	// Some servers may block request if it doesn't receive user agen
+	curl_easy_setopt(info->curl, CURLOPT_USERAGENT, info->user_agent);
```
Comment (notr1ch) [inline deps/file-updater/file-updater/file-updater.c:135]: This change seems unrelated to adding a service.

File: plugins/rtmp-services/CMakeLists.txt [hunk 4]
```diff
@@ -7,13 +7,14 @@ include_directories(${LIBCURL_INCLUDE_DIRS})
...
-	rtmp-custom.c
-	rtmp-services-main.c)
+		dacast.c
+		twitch.c
+		younow.c
+		nimotv.c
+		showroom.c
+		rtmp-common.c
+		rtmp-custom.c
+		rtmp-services-main.c)
```
Comment (notr1ch) [inline plugins/rtmp-services/CMakeLists.txt:17]: Unnecessary whitespace changes.

File: plugins/rtmp-services/CMakeLists.txt [hunk 5]
```diff
@@ -23,11 +24,12 @@ if(WIN32)
...
-	younow.h
-	nimotv.h
-	showroom.h
-	rtmp-format-ver.h)
+		dacast.h
+		twitch.h
+		younow.h
+		nimotv.h
+		showroom.h
+		rtmp-format-ver.h)
```
Comment (notr1ch) [inline plugins/rtmp-services/CMakeLists.txt:32]: Unnecessary whitespace changes.

File: plugins/rtmp-services/dacast.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+	const char *h264_profile;
+	double frame_rate;
+	int keyframe_interval;
+	const char *scanning;
+	const char *rate_control;
+	const char *audio_codec;
+	int audio_bitrate;
+	int audio_channels;
+	int audio_sample_rate;
+};
```
Comment (notr1ch) [inline plugins/rtmp-services/dacast.h:20]: There are a lot of fields here that seem unused.

File: plugins/rtmp-services/data/services.json [hunk 7]
```diff
@@ -1811,6 +1811,22 @@
                 "max video bitrate": 5000,
                 "max audio bitrate": 160
             }
+        },
+        {
+            "name": "Dacast",
+            "common": true,
```
Comment (notr1ch) [inline plugins/rtmp-services/data/services.json:1817]: Not a common service.

File: plugins/rtmp-services/dacast.c [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+		for (int i = 0; i < 3 * 100; i++) {
+			if (os_atomic_load_bool(&ingests_loaded)) {
+				break;
+			}
+			os_sleep_ms(10);
+		}
+	}
+
+	dstr_free(&uri);
+}
```
Comment (notr1ch) [inline plugins/rtmp-services/dacast.c:230]: What happens if this API request fails?

**Review by notr1ch — CHANGES_REQUESTED** [review]
In addition to the code comments, I have a lot of concerns over the design goals of this PR. If you want to dynamically control settings like bitrate, sample rate, profile, etc on a per-key basis (which is not currently possible in OBS), then a framework for doing this in a service-agnostic way is preferred, so that the code can be maintained and re-used. This would likely need to go through the RFC process - a PR like this with numerous `if (dacast)` hacks is unlikely to be merged.

**Comment by richjenks** [comment]
Hi @notr1ch thanks for reviewing this, appreciate your code comments!

The design goal is simply to make it easier to get RTMP details from the Dacast platform into OBS — the rest is secondary.

Totally understand your concerns about how the other bits are implemented. I suggest we remove them from this PR and discuss separately if that works for you.

Will respond to code comments individually 👍

File: UI/window-basic-auto-config.cpp (hunk 1 above)
Comment (richjenks) [reply UI/window-basic-auto-config.cpp:643]: We will revert this change.

File: UI/window-basic-settings-stream.cpp (hunk 2 above)
Comment (richjenks) [reply UI/window-basic-settings-stream.cpp:300]: We will revert this change.

File: deps/file-updater/file-updater/file-updater.c (hunk 3 above)
Comment (richjenks) [reply deps/file-updater/file-updater/file-updater.c:135]: Possibly related to our API, we will review.

File: plugins/rtmp-services/CMakeLists.txt (hunk 4 above)
Comment (richjenks) [reply plugins/rtmp-services/CMakeLists.txt:17]: We will revert this change.

File: plugins/rtmp-services/CMakeLists.txt (hunk 5 above)
Comment (richjenks) [reply plugins/rtmp-services/CMakeLists.txt:32]: We will revert this change.

File: plugins/rtmp-services/dacast.h (hunk 6 above)
Comment (richjenks) [reply plugins/rtmp-services/dacast.h:20]: As commented below, we'll remove this and discuss settings with you separately.

File: plugins/rtmp-services/data/services.json (hunk 7 above)
Comment (richjenks) [reply plugins/rtmp-services/data/services.json:1817]: We will change this.

File: plugins/rtmp-services/dacast.c (hunk 8 above)
Comment (richjenks) [reply plugins/rtmp-services/dacast.c:230]: I believe in this case it _should_ trigger the same error when you click **Start Streaming** as it would for an invalid key, i.e.

image

If this is acceptable please let me know so we can make this change.

**Comment by richjenks** [comment]
Hi @notr1ch , we've made the requested changes so please take a look when you can and let me know if you have any questions.

**Comment by richjenks** [comment]
Hi @notr1ch , have you had a chance to review our recent changes? We believe we've addressed all your concerns so it'd be great to know if this is now ready to be merged. Stay safe! Rich

**Comment by lmbsht** [comment]
Hello @notr1ch - I am following up on Rich's messages, is this ready to be merged? It is key for us to make sure you have everything you need here. Many thanks.
